sklearn/metrics/pairwise.py

```diff
@@ -436,6 +437,84 @@ def nan_euclidean_distances(X, Y=None, *, squared=False,
     return distances
 
 
+def _nan_fill_row_norm(r, fill_value: float = 0.):
+    p = r * r
+    p = np.nan_to_num(p, nan=fill_value)
+    return np.sum(p)
+
+
+@njit
+def _nan_fill_dot(X, Y, fill_values: np.ndarray):
+    if X.shape[1] != fill_values.shape[0]:
+        raise ValueError('X and fill_values have incompatible shapes')
+
+    p = np.zeros((X.shape[0], Y.shape[0],))
+    for i in range(0, X.shape[0]):
+        v1 = X[i, :]
+        for j in range(0, Y.shape[0]):
+            v2 = Y[j, :]
+            s = 0.
+            for k in range(0, fill_values.shape[0]):
+                if np.isnan(v1[k]) and np.isnan(v2[k]):
+                    pass
+                elif np.isnan(v1[k]) or np.isnan(v2[k]):
+                    s += fill_values[k]
+                else:
+                    s += v1[k] * v2[k]
+
+            p[i, j] = s
+
+    return p
+
+
+def nan_filled_euclidean_distances(X, fill_values, Y=None, squared=False, missing_values=np.nan, copy=True):
+    if X.dtype == np.float32 or (Y is not None and Y.dtype == np.float32):
+        raise NotImplementedError('X and Y must be np.float64 type')
+
+    X, Y = check_pairwise_arrays(X, Y, accept_sparse=False, force_all_finite='allow-nan', copy=copy)
+
+    # Get missing mask for X
+    missing_X = _get_mask(X, missing_values)
+
+    # Get missing mask for Y
+    missing_Y = missing_X if Y is X else _get_mask(Y, missing_values)
+
+    # calculate euclidean distances
+    XX = np.apply_along_axis(_nan_fill_row_norm, 1, X)
+    XX = XX[:, np.newaxis]
+
+    if X is Y and XX is not None:
+        # shortcut in the common case euclidean_distances(X, X)
+        YY = XX.T
+    else:
+        YY = np.apply_along_axis(_nan_fill_row_norm, 1, Y)
+        YY = YY[np.newaxis, :]
+
+    # if dtype is already float64, no need to chunk and upcast
+    distances = - 2 * _nan_fill_dot(X, Y, -fill_values)
+    # distances = - 2 * np.dot(X, Y.T)
+    distances += XX
+    distances += YY
+    np.maximum(distances, 0, out=distances)
```

Comment:
Should it be non-negative naturally?

Reply:
Yes, but the system might assign a very small negative number because of floating point mechanics. These lines have been copied over from the `euclidean_distances` function.